Add class to allow running depletion simulation independent of the OpenMC transport solver

yardasol opened the pull request:
This PR adds a new class, `IndependentOperator`, that enables a user to provide a dictionary of nuclides and their concentrations (in at/cm^3) and microscopic cross section data.

This allows the user to run a depletion calculation without needing to use OpenMC's transport solver.

Right now the class supports flux-based and power-based flux normalization.

Specifically, this PR adds the following:
- `deplete/independent_operator.py`
  - `IndependentOperator` class
    - A nested helper class: `_IndependentRateHelper`. This class allows us to use the code from #2115 without duplicating the `number` or `cross_sections` attributes, which saves memory.
    - A nested helper class: `_IndependentNormalizationHelper`. This class allows us to perform constant-power depletion.
  - `deplete/microxs.py`
    - A new class, `MicroXS` that stores microscopic cross section information, along with the following methods to create the class
      - `from_model`: creates one-group library via tallies from an OpenMC simulation
      - `from_csv`: reads in cross section data from a `.csv` file
      - `from_array`: constructs an object based on a numpy array.
- `tests/unit_tests`
  - `test_deplete_independent_operator.py`
  - `test_deplete_microxs.py`
- `tests/integration_tests/`
  - Move current depletion regression test to `/deplete_with_transport`
  - Create new regression test, `/deplete_no_transport` and `/microxs`
- Add some content about `IndependentOperator` and `MicroXS` to the User's guide
- Add `IndependentOperator` and `MicroXS` to the Python API docs.
- Change name of `Operator` to `CoupledOperator`, along with an alias to `Operator` and a warning about the name change on initialization.
- Change name of `TransportOperator` to `DepletionOperator` to reflect the more generic use case for such an operator.
- Updates the code in the Chapter 8 of the user's guide to reflect the new `Operator` API (using `model` instead of `geometry` and `settings`)

Additionally, much of the code between `Operator` and `IndependentOperator` can be shared. This PR refactors this shared code into a a new class, `OpenMCOperator`, as a subclass of `TransportOperator`

Specifically, this PR does the following:
- Adds a new class `OpenMCOperator`. This class has the following abstract functions to be implemented by subclasses:
  - `_differentiate_burnable_mats()`
  - `_load_previous_results()`
  - `_get_nuclides_with_data()`
  - `_get_helper_classes()`
  - `_update_materials()`
  - `__call__()` (inherited from `TransportOperator`)
  - `write_bos_data()` (inherited from `TransportOperator`)
- Moves the following functions from `Operator` to `OpenMCOperator`:
  - `_distribute()`
  - `_get_burnable_mats()`
  - `_extract_number()`
  - `_set_number_from_mat()`
  - `_set_number_from_results()`
  - `Operator._get_tally_nuclides()` --> `OpenMCOperator._get_reaction_nuclides()`
  - `Operator._unpack_tallies_and_normalize()` --> `OpenMCOperator._calculate_reaction_rates()`
  - `get_results_info()`

- Splits the following functions into into reusable and non-reusable parts. The non-resuable parts stay in the original function under `Operator`, and the reusable parts are moved to a function of the same name (unless otherwise noted) under `OpenMCOperator`.
  - `__init__()`. The `OpenMCOperator.__init__()` implementation utilizes the following abstract functions:
    - `_differentiate_burnable_mats()`
    - `_load_previous_results()`
    - `_get_nuclides_with_data()`
    - `_get_helper_classes()`
  - `initial_condition()`
  - `__call__()` (the reusable part moved to `OpenMCOperator` is named `_update_materials_and_nuclides()`)
- Adds new section in the API docs for the `deplete` module on intermediate classes -- classes that implement abstract base classes and hold shared methods for their subclasses
  - `TalliedFissionYieldHelper`
  - `OpenMCOperator`

yardasol commented:
Added some basic unit tests, nothing major.

Working on getting an integration test up and running. Need to generate the 1-group cross sections first.

A detail I've notices that should be addressed (in the future), is that in practice, a user would need to update the microscopic cross sections each iteration as the flux spectrum as the fuel composition evolves, especially if we add support for multigroup microscopic xs data.

paulromano reviewed (changes requested):
This is a great start @yardasol! In addition to the line comments, one thing I notice is that a lot of the code looks very similar to what already exists in operator.py. We should think about whether it's possible to reduce some of the redundant code by moving it into a base class or otherwise combining.

Before getting to the point of merging, we should also consider how this will extend to multiple materials as well as using multigroup cross sections/flux instead of one group.

yardasol commented on docs/source/usersguide/depletion.rst:
```diff
@@ -177,3 +176,39 @@ across all material instances.
+data arrays::
+
+    ...
+    micro_xs = FluxDepletionOperator.create_micro_xs_from_csv(micro_xs_path)
```
I'm just trying to demonstrate how one would initiate a `FluxDepletionOperator` object. I have some discussion on the structure of the `micro_xs` parameter in the docstrings for that class, but I can try to add some details about that in the guide if you think it'd be useful.

yardasol commented on openmc/deplete/flux_operator.py:
```diff
@@ -0,0 +1,529 @@
+        return [nuc for nuc in nuc_list if nuc in self.chain]
+
+    def _get_all_nuclides_in_simulation(self):
+        """Determine nuclides that will show up in the simulation.
```
Correct! I'll make this more explicit.

shimwell commented:
I'm just taking a look at this branch as I am keen to see if it will reduce the amount of compute I'm currently using for depletion.

Currently I am using the  ```CF4Integrator``` depletion operator with a few time steps and zero source rates in several timesteps.

```python
integrator = openmc.deplete.CF4Integrator(
        operator=operator,
        timesteps=[60, 60, 60, 60],
        source_rates=[1e21, 0, 0, 0],
    )
```

This appears to run **4** openmc transport simulations when I was hoping to get away with just one transport simulation as the source rate is 0 for the last 3 time steps.

I'm wondering if the addition of the ```FluxDepletionOperator``` in this PR will help me perform such a depletion with just 1 transport step?

shimwell commented:
I'm seeing transport being run on time each step.

I can see the ```if source_rate == 0.0:``` in the operator.py file but I notice the ```openmc.lib.run()``` comes a few lines later and is outside of this if statement.

This is perhas a bit unrelated to your pull request but I'm wondering if we can move the ```openmc.lib.run()``` into an if statement block that checks for zero flux as well

paulromano commented:
@shimwell The `CFIntegrator` requires four transport solves _per timestep_ (see mathematical definition [here](https://docs.openmc.org/en/latest/pythonapi/generated/openmc.deplete.CF4Integrator.html#openmc.deplete.CF4Integrator)), so the fact that you are seeing only four solves (from the first timestep) indicates that on the subsequent three timesteps there are no transport solves happening.

yardasol commented:
Some of the test runs keep failing during the `after success` step:
```
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.8.13/x64/lib/python3.8/site-packages/coveralls/cli.py", line 95, in main
    result = coverallz.wear()
  File "/opt/hostedtoolcache/Python/3.8.13/x64/lib/python3.8/site-packages/coveralls/api.py", line 257, in wear
    return self.submit_report(json_string)
  File "/opt/hostedtoolcache/Python/3.8.13/x64/lib/python3.8/site-packages/coveralls/api.py", line 293, in submit_report
    raise CoverallsException(
coveralls.exception.CoverallsException: Could not submit coverage: 502 Server Error: Bad Gateway for url: https://coveralls.io/api/v1/jobs
Error: Process completed with exit code 1.
```

paulromano commented:
@yardasol Don't worry about coveralls errors -- that seems to be affecting OpenMC as a whole and is not related to anything on your branch

paulromano commented on docs/source/usersguide/depletion.rst:
```diff
@@ -113,21 +112,24 @@ should be, including indirect components. Some examples are provided below::
     fission_q = {"U235": 202e+6}  # energy in eV
 
+    # create a Model object
+    model  = openmc.model.Model(geometry, settings)
```
```suggestion
    model  = openmc.Model(geometry, settings)
```

paulromano commented on docs/source/usersguide/depletion.rst:
```diff
@@ -177,3 +179,45 @@ across all material instances.
+transport solver using the :class:`FluxDepletionOperator` class. This class
+has two ways to initalize it; the default constructor accepts an 
+:class:`openmc.Materials` object, a flux spectra, and one-group microscopic 
+cross sections as a :class:`pandas.Dataframe`, while the `from_nuclides` 
```
```suggestion
cross sections as a :class:`pandas.DataFrame`, while the `from_nuclides` 
```

paulromano commented on docs/source/usersguide/depletion.rst:
```diff
@@ -177,3 +179,45 @@ across all material instances.
     number of tallies and material definitions.
 
+Transport-independent depletion
+-------------------------------
```
```suggestion
===============================
```

paulromano commented on docs/source/usersguide/depletion.rst:
```diff
@@ -177,3 +179,45 @@ across all material instances.
+   possible and likely in the near future.
+
+OpenMC supports running depletion calculations independent of the OpenMC
+transport solver using the :class:`FluxDepletionOperator` class. This class
```
```suggestion
transport solver using the :class:`~openmc.deplete.FluxDepletionOperator` class. This class
```

paulromano commented on openmc/deplete/flux_operator.py:
```diff
@@ -0,0 +1,464 @@
+"""Pure depletion operator
+
+This module implements a pure depletion operator that uses user provided fluxes
```
```suggestion
This module implements a pure depletion operator that uses user-provided fluxes
```

paulromano commented on openmc/deplete/openmc_operator.py:
```diff
@@ -0,0 +1,567 @@
+    diff_burnable_mats : bool, optional
+        Whether to differentiate burnable materials with multiple instances.
+        Volumes are divided equally from the original material volume.
+        Default: False.
```
Defaults are listed in the function signature so not really necessary to put them in the docstring
```suggestion
```

paulromano commented on openmc/deplete/openmc_operator.py:
```diff
@@ -0,0 +1,567 @@
+        Only done for nuclides with reaction rates.
+        Defaults to 1.0e3.
+    helper_kwargs : dict
+        Arguments for helper classes
```
```suggestion
        Keyword arguments for helper classes
```

paulromano commented on openmc/deplete/openmc_operator.py:
```diff
@@ -0,0 +1,567 @@
+            Results from a previous depletion calculation
+
+        """
+        self.number = AtomNumber(
+            local_mats, all_nuclides, volume, len(
+                self.chain))
```
```suggestion
        self.number = AtomNumber(local_mats, all_nuclides, volume, len(self.chain))
```

paulromano commented on openmc/deplete/openmc_operator.py:
```diff
@@ -0,0 +1,567 @@
+
+        Parameters
+        ----------
+        reaction_rate_mode : str
+            Indicates the subclass of :class:`ReactionRateHelper` to
+            instantiate.
+        normalization_mode : str
+            Indicates the subclass of :class:`NormalizationHelper` to
+            instantiate.
+        fission_yield_mode : str
+            Indicates the subclass of :class:`FissionYieldHelper` to instatiate.
+        reaction_rate_opts : dict
+            Keyword arguments that are passed to the :class:`ReactionRateHelper`
+            subclass.
+        fission_yield_opts : dict
+            Keyword arguments that are passed to the :class:`FissionYieldHelper`
+            subclass.
```
This doesn't match what arguments are passed (only `helper_kwargs`)

yardasol replied:
I've removed this. It was cruft which I forgot to move.

paulromano reviewed (changes requested):
Really nice work on this @yardasol! Here's a first round of comments for you:

yardasol commented on openmc/deplete/flux_operator.py:
```diff
@@ -0,0 +1,464 @@
+        """
+        return set(cross_sections.index)
+
+    class FluxTimesXSHelper(ReactionRateHelper):
```
I renamed it to `_FluxDepletionRateHelper`

yardasol commented on openmc/deplete/flux_operator.py:
```diff
@@ -0,0 +1,464 @@
+        nuc_ind_map = {ind: nuc for nuc, ind in rates.index_nuc.items()}
+        rxn_ind_map = {ind: rxn for rxn, ind in rates.index_rx.items()}
+
+        self._rate_helper = self.FluxTimesXSHelper(
+            self.reaction_rates.n_nuc, self.reaction_rates.n_react, self)
+        self._rate_helper.nuc_ind_map = nuc_ind_map
+        self._rate_helper.rxn_ind_map = rxn_ind_map
```
Good catch!

yardasol commented on openmc/deplete/operator.py:
```diff
@@ -244,65 +224,126 @@ def __init__(self, model, chain_file=None, prev_results=None,
+
+        Parameters
+        ----------
+        **kwargs : ...
+        reaction_rate_mode : str
+            Indicates the subclass of :class:`ReactionRateHelper` to
+            instantiate.
+        normalization_mode : str
+            Indicates the subclass of :class:`NormalizationHelper` to
+            instatiate.
+        fission_yield_mode : str
+            Indicates the subclass of :class:`FissionYieldHelper` to instantiate.
+        reaction_rate_opts : dict
+            Keyword arguments that are passed to the :class:`ReactionRateHelper`
+            subclass.
+        fission_yield_opts : dict
+            Keyword arguments that are passed to the :class:`FissionYieldHelper`
+            subclass.
```
Done

yardasol commented on openmc/deplete/flux_operator.py:
```diff
@@ -0,0 +1,464 @@
+
+        Parameters
+        ----------
+        outer : openmc.deplete.FluxDepletionOperator
+            Reference to the object encapsulate FluxTimesXSHelper.
+            We pass this so we don't have to duplicate the ``number`` object.
+        n_nucs : int
+            Number of burnable nuclides tracked by :class:`openmc.deplete.Operator`
+        n_react : int
+            Number of reactions tracked by :class:`openmc.deplete.Operator`
```
Done

yardasol commented on openmc/deplete/flux_operator.py:
```diff
@@ -0,0 +1,464 @@
+    micro_xs : pandas.DataFrame
+        DataFrame with nuclides names as index and microscopic cross section
+        data in the columns. Cross section units are [cm^-2].
+    flux_spectra : float
+        Flux spectrum [n cm^-2 s^-1]
```
Done

yardasol commented on openmc/deplete/flux_operator.py:
```diff
@@ -0,0 +1,464 @@
+                            # values.
+                            if val < -1.0e-21:
+                                print(
+                                    "WARNING: nuclide ",
+                                    nuc,
+                                    " in material ",
+                                    mat,
+                                    " is negative (density = ",
+                                    val,
+                                    " at/barn-cm)")
```
Done

yardasol commented on openmc/deplete/flux_operator.py:
```diff
@@ -0,0 +1,464 @@
+        self._rate_helper.nuc_ind_map = nuc_ind_map
+        self._rate_helper.rxn_ind_map = rxn_ind_map
+
+        self._normalization_helper = SourceRateHelper()
```
As we discussed, I'm going to add an addtional option to use `ChainFissionHelper`

paulromano commented on openmc/deplete/abc.py:
```diff
@@ -523,8 +525,10 @@ class Integrator(ABC):
         Power density of the reactor in [W/gHM]. It is multiplied by
         initial heavy metal inventory to get total power if ``power``
         is not speficied.
+    flux : float or iterable of float, optional
+        neutron flux in [neut/s-cm^2] for each interval in :attr:`timesteps`
```
Since this is being treated exactly the same as `source_rates`, I'm not sure we should introduce a new argument just for this. We can expand the description of `source_rates` below to say that it can also represent the neutron flux (which sort of _is_ the source rate in a sense) and then also make it clear in the .rst documentation.

yardasol replied:
I agree I think we can just add more spec to `source_rates`. Where in the .rst documentation are you referring to?

paulromano commented on docs/source/usersguide/depletion.rst:
```diff
@@ -19,14 +19,13 @@ transmutation equations and the method used for advancing time. At present, the
 
-    geom = openmc.Geometry()
-    settings = openmc.Settings()
+    model = openmc.model.Model()
```
```suggestion
    model = openmc.Model()
```

paulromano commented on docs/source/usersguide/depletion.rst:
```diff
@@ -19,14 +19,13 @@ transmutation equations and the method used for advancing time. At present, the
 implemented and no changes to the depletion solver itself would be needed. The
-operator class requires a :class:`openmc.Geometry` instance and a
-:class:`openmc.Settings` instance::
+operator class requires a :class:`openmc.model.Model` instance containing 
```
```suggestion
operator class requires a :class:`~openmc.Model` instance containing 
```

paulromano commented on openmc/deplete/openmc_operator.py:
```diff
@@ -0,0 +1,569 @@
+    cross_sections : str or pandas.DataFrame
+            Path to continuous energy cross section library, or object
+            containing one-group cross-sections.
+
```
```suggestion
```

paulromano commented on openmc/deplete/openmc_operator.py:
```diff
@@ -0,0 +1,569 @@
+
+            # Accumulate energy from fission
+            if fission_ind is not None:
+                self._normalization_helper.update(tally_rates[:, fission_ind], mat_index=mat_index)
```
Need some line wrapping here

paulromano commented on openmc/deplete/operator.py:
```diff
@@ -244,173 +222,32 @@ def __init__(self, model, chain_file=None, prev_results=None,
-        openmc.lib.statepoint_write(
-            "openmc_simulation_n{}.h5".format(step),
-            write_source=False)
+        helper_kwargs = dict()
```
```suggestion
```

paulromano commented on openmc/deplete/operator.py:
```diff
@@ -434,137 +271,104 @@ def _differentiate_burnable_mats(self):
-            self.number.set_atom_density(mat_id, nuclide, atom_per_cc)
+        # Select and create fission yield helper
+        fission_helper = self._fission_helpers[fission_yield_mode]
+        if fission_yield_opts is None:
+            fission_yield_opts = {}
```
```suggestion
```

yardasol replied:
The same comment applies as it does to this `if` statement in `FluxDepletionOperator._get_helper_classes()`

paulromano commented on openmc/deplete/operator.py:
```diff
@@ -720,103 +431,90 @@ def _unpack_tallies_and_normalize(self, source_rate):
-            for name in library['materials']:
-                if name not in nuclides:
-                    nuclides.add(name)
+                                      ' at/barn-cm)')
```
To be consistent with usage elsewhere in the API
```suggestion
                                      ' atom/b-cm)')
```

paulromano commented on openmc/deplete/flux_operator.py:
```diff
@@ -0,0 +1,594 @@
+                                print(f'WARNING: nuclide {nuc} in material'
+                                      f'{mat} is negative (density = {val}'
+
+                                      ' at/barn-cm)')
```
For consistency
```suggestion
                                      ' atom/b-cm)')
```

paulromano commented on tests/regression_tests/deplete_no_transport/test.py:
```diff
@@ -0,0 +1,115 @@
+    fuel.add_element("U", 1, percent_type="ao", enrichment=4.25)
+    fuel.add_element("O", 2)
+    fuel.set_density("g/cc", 10.4)
+    fuel.depletable=True
```
```suggestion
    fuel.depletable = True
```

paulromano commented on tests/regression_tests/deplete_no_transport/test.py:
```diff
@@ -0,0 +1,115 @@
+
+    if from_nuclides:
+        nuclides = {}
+        for nuc, dens in fuel.get_nuclide_atom_densities().items():
+            nuclides[nuc] = dens * 1e24
```
```suggestion
        for nuc, atom_per_b_cm in fuel.get_nuclide_atom_densities().items():
            nuclides[nuc] = atom_per_b_cm * 1e24
```

paulromano replied:
This also makes me wonder if we should just have `from_nuclides` take densities in atom/b-cm rather than atom/cm³

yardasol replied:
I think it's normal for people to have nuclide concentrations in units other than atom/b-cm. Perhaps we should add a unit specifier to `from_nuclides`?

yardasol replied:
I've added such a specifier

paulromano commented on tests/regression_tests/deplete_no_transport/test.py:
```diff
@@ -0,0 +1,115 @@
+""" Transport-free depletion test suite """
+
+from math import floor
+import shutil
+from pathlib import Path
+
+from difflib import unified_diff
+import numpy as np
+import pytest
+import openmc
+from openmc.data import JOULE_PER_EV
+import openmc.deplete
+from openmc.deplete import FluxDepletionOperator
+
+
+from tests.regression_tests import config
```
```suggestion
from math import floor
import shutil
from pathlib import Path
from difflib import unified_diff

import numpy as np
import pytest
import openmc
from openmc.data import JOULE_PER_EV
import openmc.deplete
from openmc.deplete import FluxDepletionOperator

from tests.regression_tests import config
```

yardasol commented on openmc/deplete/flux_operator.py:
```diff
@@ -0,0 +1,594 @@
+        keff : 2-tuple of float, optional
+           keff eigenvalue and uncertainty from transport calculation.
+           Default is None.
+        normalization_mode : {"constant-power", "constant-flux"}
```
So after skimming that paper, I have a few points of understanding I'd like to reach consensus on:
- the `constant-power` normalization mode _is_ constant power but with the assumptions that the flux is the same at the beginning of the depletion step as at the end of the depletion step, which will give a flux that is different from the "true" value, which may not be a very good assumption.
- to get better results for the flux calculated from the constant power, we need to perform substeps to converge to a value of flux that is closer to the "true" value
- Because we can only accept a single `DataFrame` for the one-group cross-sections, even if we wanted to do this kind of substep renormalization, we couldn't because that would require us to calculate new cross sections for each substep.

We keep running into this issue of having to recalculate the cross sections. Implementing this kind of functionality (and additionally all the extra cool things you can do with it, like this substep renormalization) would be an interesting project!

I think the `constant-fission` name is fine, but we probably should try to be consistent with the rest of the `deplete` module and revert it to `source-rate`.

paulromano replied:
> * with the assumptions that the flux is the same at the beginning of the depletion step as at the end of the depletion step

That _is_ the constant flux assumption :smile: The only difference really is how you got that constant flux in the first place — specifying it explicitly or deducing it from the power.

> * we need to perform substeps to converge to a value of flux

Yes, that is the only way I could see to adjust the flux so that the power is not decreasing over the full step.

> * Because we can only accept a single `DataFrame` for the one-group cross-sections, even if we wanted to do this kind of substep renormalization, we couldn't

Cross sections are only one piece of the puzzle. We _would_ have updated number densities, so even if the cross sections are constant, you need a higher flux to account for the decrease in the density of fissionable nuclides to maintain constant power. I think that's actually the primary effect -- I wouldn't expect the micros to change much over a single step since the spectrum is probably about the same.

> Implementing this kind of functionality ... would be an interesting project!

Agreed!

yardasol commented on openmc/deplete/flux_operator.py:
```diff
@@ -0,0 +1,594 @@
+
+        """
+            volume = self._op.number.get_mat_volume(mat_index)
+            densities = np.empty(np.shape(fission_rates))
```
handy!

yardasol commented on openmc/deplete/flux_operator.py:
```diff
@@ -0,0 +1,594 @@
+        # on this process
+        if comm.size != 1:
+            prev_results = self.prev_res
+            self.prev_res = Results()
+            mat_indexes = _distribute(range(len(self.burnable_mats)))
```
I added an import statement to fix this

yardasol commented on openmc/deplete/flux_operator.py:
```diff
@@ -0,0 +1,594 @@
+
+        """
+
+        def __init__(self, n_nuc, n_react, op):
+            super().__init__(n_nuc, n_react)
+            rates = op.reaction_rates
```
Hadn't considered that. Good suggestion!

yardasol commented on openmc/deplete/flux_operator.py:
```diff
@@ -0,0 +1,594 @@
+
+        # Select and create fission yield helper
+        fission_helper = ConstantFissionYieldHelper
+        if fission_yield_opts is None:
+            fission_yield_opts = {}
```
This suggestion won't work with the current structure of the code, which is why I didn't incorporate it from your previous round of review, but perhaps I should have said something about this.

In `__init__`, we have
```python
def __init__(self,
                  materials,
                  micro_xs,
                  chain_file,
                  keff=None,
                  normalization_mode='constant-flux',
                  fission_q=None,
                  prev_results=None,
                  reduce_chain=False,
                  reduce_chain_level=None,
                  fission_yield_opts=None):
         .
         .
         .
         helper_kwargs = {'normalization_mode': normalization_mode,
                          'fission_yield_opts': fission_yield_opts}
         .
         .
         .
``` 

The default value of `fission_yield_opts` is `None`, and since it is always passed to `helper_kwargs`, the `helper_kwargs.get('fission_yield_opts', {})` fuction inside of `_get_helper_classes` will always return the value of `fission_yield_opts` stored in `helper_kwargs`, and never return the alternate value given (in this case, an empty `dict`). So we need the `if` statement on 373 to avoid an error from trying to pass a `None` value to `fission_helper`.

If you can think of a good way to restructure this so we don't need the `if` statement, please make a suggestion!

paulromano replied:
Ah, right! In that case, let's just move the `if` block further up, i.e.
```Python
if fission_yield_opts is None:
    fission_yield_opts = {}
helper_kwargs = ...
```

yardasol commented on openmc/deplete/flux_operator.py:
```diff
@@ -0,0 +1,594 @@
+        vec : list of numpy.ndarray
+            Total atoms to be used in function.
+        source_rate : float
+            Power in [W] or source rate in [neutron/sec]
```
I've replaced `source rate` with flux and changed the units, because passing a source rate to this class isn't quite correct.

Should I add a warning statement to the `Integrator` class to alert a user that they may get unexpected results if passing a `source_rate` parameter instead of `flux` or `power` if the `operator` is a `FluxDepletionOperator` object?

yardasol commented on openmc/deplete/flux_operator.py:
```diff
@@ -0,0 +1,594 @@
+        data : ndarray of floats
+            Array containing one-group microscopic cross section information for each
+            nuclide and reaction.
+        units : {'barn', 'cm^2'}, optional
```
Hmm, probably not. that would actually be quite ugly now that I think about it. I'll remove this.

yardasol commented on openmc/deplete/flux_operator.py:
```diff
@@ -0,0 +1,594 @@
+
+
+    @staticmethod
+    def generate_1g_cross_sections(model, reaction_domain, reactions=['(n,gamma)', '(n,2n)', '(n,p)', '(n,a)', '(n,3n)', '(n,4n)', 'fission'],  energy_bounds=(0, 20e6), write_to_csv=True, filename='micro_xs.csv'):
```
I'm actually glad you think so because I was not sure where to put it when I wrote it.

yardasol commented on openmc/deplete/flux_operator.py:
```diff
@@ -0,0 +1,594 @@
+    @staticmethod
+    def generate_1g_cross_sections(model, reaction_domain, reactions=['(n,gamma)', '(n,2n)', '(n,p)', '(n,a)', '(n,3n)', '(n,4n)', 'fission'],  energy_bounds=(0, 20e6), write_to_csv=True, filename='micro_xs.csv'):
+        """Helper function to generate a one-group cross-section dataframe
+        using OpenMC. Note that the ``openmc`` C executable must be compiled.
```
I keep thinking that OpenMC is written in C and not C++...

yardasol commented on openmc/deplete/flux_operator.py:
```diff
@@ -0,0 +1,594 @@
+                xs[rxn] = ArbitraryXS(rxn, domain=reaction_domain, groups=groups, by_nuclide=True)
+            tallies += xs[rxn].tallies.values()
+
+        model.tallies = tallies
```
Is this so, if a user is running in a notebook, they can rerun without needing to reinitialize their `model` object?

yardasol replied:
I like option 2

yardasol commented on openmc/deplete/flux_operator.py:
```diff
@@ -0,0 +1,594 @@
+            df.rename({'mean':rxn}, axis=1, inplace=True)
+            micro_xs = pd.concat([micro_xs, df], axis=1)
+
+        if write_to_csv:
+            micro_xs.to_csv(filename)
+            return None
+        else:
+            return micro_xs
```
I also thought this should be the case when I looked at it again today.

yardasol commented on openmc/deplete/flux_operator.py:
```diff
@@ -0,0 +1,594 @@
+            Indicate how reaction rates should be calculated.
+            ``"constant-power"`` uses the fission Q values from the depletion
+            chain to compute the flux based on the power. ``"constant-flux"``
+            uses the value stored in `_normalization_helper` as the flux.
```
Fixed it

yardasol commented:
@paulromano I think I addressed your second round of comments. I've left suggestions that I'd like feedback on my solution to unresolved.

paulromano commented on openmc/deplete/flux_operator.py:
```diff
@@ -0,0 +1,601 @@
+
+    return micro_xs
+
+class FluxDepletionOperator(OpenMCOperator):
```
I'm reconsidering whether this name is apt. What do you think of `IndependentOperator`? I think the "Depletion" part is probably not necessary since it's already namespaced in `openmc.deplete`.

paulromano replied:
Another alternative: `ExternalOperator` (i.e., transport is run externally)

yardasol replied:
I agree. It's fine as a preliminary name but we should pick something more polished before this gets merged.

yardasol replied:
I think `ExternalOperator` makes the most sense

paulromano commented on openmc/deplete/flux_operator.py:
```diff
@@ -0,0 +1,601 @@
+        if ``"normalization_mode" == "fission-q"``.
+    reduce_chain : bool, optional
+        If True, use :meth:`openmc.deplete.Chain.reduce` to reduce the
+        depletion chain up to ``reduce_chain_level``. Default is False.
```
```suggestion
        depletion chain up to ``reduce_chain_level``.
```

paulromano commented on openmc/deplete/flux_operator.py:
```diff
@@ -0,0 +1,601 @@
+        return set(cross_sections.index)
+
+    class _FluxDepletionNormalizationHelper(ChainFissionHelper):
+        """Class for calculating one-group flux
+        based on a power.
```
```suggestion
        """Class for calculating one-group flux based on a power.
```

paulromano commented on openmc/deplete/flux_operator.py:
```diff
@@ -0,0 +1,601 @@
+                            number_i[mat, nuc] = 0.0
+
+    @staticmethod
+    def create_micro_xs_from_data_array(nuclides, reactions, data):
```
So now we have:
- `generate_1g_cross_sections`
- `FluxDepletionOperator.create_micro_xs_from_data_array`
- `FluxDepletionOperator.create_micro_xs_from_csv`

I'm wondering if maybe introducing a class to store 1g cross sections (that subclasses pandas.DataFrame) might be worthwhile. In that case, we could have:
```Python
class MicroXS(pandas.DataFrame):
    @classmethod
    def from_model(cls, ...):
        ...

    @classmethod
    def from_array(cls, ...):
        ...

    @classmethod
    def from_csv(cls, ...):
        ...
```
What do you think? I could also envision other uses of this class in the future, e.g., interpolating between multiple sets of micros.

yardasol replied:
I like this idea. I'll add it to the PR. Any idea on where it should go? Probably not in `deplete`...

paulromano replied:
I'd say for now, you can leave it in `deplete` since that's the only immediate application. We can always reconsider where the best place is for it to live later

paulromano reviewed (changes requested):
A few more suggestions and thoughts:

yardasol commented on openmc/deplete/flux_operator.py:
```diff
@@ -0,0 +1,601 @@
+        tallies += xs[rxn].tallies.values()
+
+    model.tallies = tallies
+    statepoint_path = model.run()
```
Great idea.

yardasol commented on openmc/deplete/flux_operator.py:
```diff
@@ -0,0 +1,601 @@
+        df.rename({'mean':rxn}, axis=1, inplace=True)
+        micro_xs = pd.concat([micro_xs, df], axis=1)
+
+    if write_to_csv:
+        micro_xs.to_csv(filename)
```
I think that makes sense.

yardasol commented on docs/source/usersguide/depletion.rst:
```diff
@@ -177,3 +179,57 @@ across all material instances.
+
+.. important::
+
+   Make sure you set the correct parameter in the :class:`openmc.abc.Integrator` class. Use the ``source_rates`` parameter when ``normalization_mode == source-rate``, and use ``power`` or ``power_density`` when ``normalization_mode == fission-q``.
+
+.. warning::
+
+   The accuracy of results when using ``fission-q`` is entirely dependent on your depletion chain. Make sure it has sufficient data to resolve the dynamics of your particular scenario. 
```
Sure thing

yardasol commented:
@paulromano thoughts on changing `Operator` to `CoupledOperator`?

yardasol commented:
Also, let's wait for #2139 before merging so we can add that functionality to the `from_nuclides` method.

yardasol commented:
>@paulromano thoughts on changing Operator to CoupledOperator and TransportOperator to Operator? This might clear up the purpose of each with the introduction of IndependentOperator into the mix. We should also update some of the terminology in the user's guide and API docpages to reflect the expanded scope of Operator (previously called TransportOperator)

I discussed this point in private messages with @paulromano. It was brought up that changing `TransportOperator` to `Operator` would completely break backwards compatibility. I agree that we do not want to do this. I still want to change the name of `TransportOperator`, perhaps `GenericOperator` or `ReactionRateOperator`?

I still also think we should change `Operator` to `CoupledOperator`, but we should probably include an alias so that a user can still use `Operator` to initalize a class, and perhaps a deprecation warning as well?

paulromano commented on openmc/deplete/coupled_operator.py:
```diff
@@ -0,0 +1,529 @@
+                 reaction_rate_mode="direct", reaction_rate_opts=None,
+                 reduce_chain=False, reduce_chain_level=None):
+        # warn of name change
+        warn(
+            "The Operator(...) class has been renamed and will "
+            "be removed in a future version of OpenMC. Use "
+            "CoupledOperator(...) instead.",
+            FutureWarning
+        )
```
This warning should only be displayed if the user actually calls `Operator`. An easy way to do this is:
```Python
def Operator(*args, **kwargs):
    warn(...)
    return CoupledOperator(*args, **kwargs)
```

paulromano reviewed (changes requested):
Thanks for the updates! Here are a few more comments

yardasol commented on openmc/deplete/microxs.py:
```diff
@@ -0,0 +1,178 @@
+        data : ndarray of floats
+            Array containing one-group microscopic cross section values for
+            each nuclide and reaction
+        units : {'b', 'cm^2'}
+            Units of the cross section values in ``data``
```
I added it back in because the `from_arrays` and `from_csv` methods have no way to actually check the units of the data. But I can remove it if you think it isn't useful.

yardasol commented:
@paulromano thanks for the comments! Any feedback on my updates to the docpages?

paulromano commented on openmc/deplete/abc.py:
```diff
@@ -76,14 +75,15 @@ def change_directory(output_dir):
 
-class TransportOperator(ABC):
-    """Abstract class defining a transport operator
+class DepletionOperator(ABC):
```
I think `TransportOperator` more accurately describes what this series of classes is all about: given a composition, return k<sub>eff</sub> and reaction rates. Obviously the new `IndependentOperator` doesn't do any transport, but it also doesn't do any depletion. `ReactionRateOperator` could be more apt, although it doesn't reflect the k<sub>eff</sub> part...

yardasol replied:
Okay, I'll switch it back to TransportOperator then.

paulromano commented on docs/source/pythonapi/deplete.rst:
```diff
@@ -12,19 +12,20 @@ Primary API
-algorithms for depletion calculations, which are described in detail in Colin
-Josey's thesis, `Development and analysis of high order neutron
-transport-depletion coupling algorithms <http://hdl.handle.net/1721.1/113721>`_.
+The former is responsible for calcuating retaining important information required for depletion. The most common examples are reaction rates and power
```
```suggestion
The former is responsible for calculating and retaining important information required for depletion. The most common examples are reaction rates and power
```

paulromano commented on docs/source/pythonapi/deplete.rst:
```diff
@@ -189,24 +194,57 @@ total system energy.
    helpers.FissionYieldCutoffHelper
    helpers.FluxCollapseHelper
 
+The :class:`openmc.deplete.IndependentOperator` uses inner class subclassed from
+those listed to perform similar calculations.
```
```suggestion
The :class:`openmc.deplete.IndependentOperator` uses inner classes subclassed from
those listed above to perform similar calculations.
```

paulromano commented on docs/source/usersguide/depletion.rst:
```diff
@@ -4,50 +4,30 @@
+burnup, calculations through the :mod:`openmc.deplete` Python module. OpenMC
+uses transmutation reaction rates to solve a set of transmutation equations
+that determine the evolution of nuclide densities within a material. The
+nuclide densities predicted as some future time are then used to determine
```
```suggestion
nuclide densities predicted at some future time are then used to determine
```

paulromano commented on docs/source/usersguide/depletion.rst:
```diff
@@ -177,3 +192,163 @@ across all material instances.
+
+.. note::
+   
+   This feature is still under heavy development and has yet to be verifed
+   code-to-code . API changes and feature additions are possible and likely in
+   the near future.
```
```suggestion
   This feature is still under heavy development and has yet to be rigorously
   verified. API changes and feature additions are possible and likely in
   the near future.
```

paulromano commented on docs/source/usersguide/depletion.rst:
```diff
@@ -177,3 +192,163 @@ across all material instances.
+                     [0.01, 0.5]])
+    micro_xs = MicroXS.from_array(nuclides, reactions, data)
+
+.. important ::
```
```suggestion
.. important::
```

paulromano commented on docs/source/usersguide/depletion.rst:
```diff
@@ -177,3 +192,163 @@ across all material instances.
+   ``normalization_mode == source-rate``, and use ``power`` or ``power_density``
+   when ``normalization_mode == fission-q``.
+
+1. ``soure-rate`` normalization, which assumes the ``source-rate`` provided by
```
```suggestion
1. ``source-rate`` normalization, which assumes the ``source-rate`` provided by
```

paulromano commented on docs/source/usersguide/depletion.rst:
```diff
@@ -177,3 +192,163 @@ across all material instances.
+1. ``soure-rate`` normalization, which assumes the ``source-rate`` provided by
+   the time integrator is a flux, and obtains the reaction rates by multiplying
+   the cross-sections by the ``source-rate``.
+2. ``fission-q`` normalization, which assumes the ``source-rate`` provided by
```
In this case, doesn't the user pass `power` or `power_density`, not `source-rate`?

yardasol replied:
Correct. I think I was referring to how internally in `Integrator`, the variable passed to the operator classes no matter what the user provides is `source_rate`, but that doesn't really make a difference to the user. I'll adjust the wording here.

paulromano commented on docs/source/usersguide/depletion.rst:
```diff
@@ -177,3 +192,163 @@ across all material instances.
+   However, there is a method to converge to a more accurate value for flux by
+   using substeps during time integration.
+   `This paper <https://doi.org/10.1016/j.anucene.2016.05.031>`_ provides a
+   good discussion of this method. Hopefully such a method will be implemented
+   in OpenMC in the near future.
```
I try to avoid making promises in the documentation
```suggestion
   good discussion of this method.
```

yardasol replied:
haha that's a good recommendation

paulromano commented on docs/source/usersguide/depletion.rst:
```diff
@@ -177,3 +192,163 @@ across all material instances.
+different cases of a particular scenario. However, running a depletion
+simulation with multiple materials using the ``fission-q`` normalization method
+treats each material as part of the same "reactor" due to how ``fission-q``
+normalization conglomerates energy values from each material to a single value.
```
```suggestion
normalization accumulates energy values from each material to a single value.
```

paulromano commented on openmc/deplete/microxs.py:
```diff
@@ -0,0 +1,169 @@
+            df.rename({'mean': rxn}, axis=1, inplace=True)
+            micro_xs = concat([micro_xs, df], axis=1)
+
+        micro_xs._units = 'b'
+
```
```suggestion
```

paulromano commented on docs/source/pythonapi/deplete.rst:
```diff
@@ -12,19 +12,20 @@ Primary API
-algorithms for depletion calculations, which are described in detail in Colin
-Josey's thesis, `Development and analysis of high order neutron
-transport-depletion coupling algorithms <http://hdl.handle.net/1721.1/113721>`_.
+The former is responsible for calcuating and retaining important information required for depletion. The most common examples are reaction rates and power
```
```suggestion
The former is responsible for calculating and retaining important information
required for depletion. The most common examples are reaction rates and power
```

paulromano commented on docs/source/methods/depletion.rst:
```diff
@@ -103,16 +103,16 @@ integrate over the entire timestep.
+to compute a transport solution, i.e., to evaluate :math:`\mathbf{A}` for a
+given :math:`\mathbf{n}`. Thus, the cost of a method scales with the number of
+:math:`\mathbf{A}` evaluations that are performed per timestep. On the other
+hand, methods that require more evaluations generally achieve higher accuracy. The predictor method only requires one evaluation and its error converges as
```
```suggestion
hand, methods that require more evaluations generally achieve higher accuracy.
The predictor method only requires one evaluation and its error converges as
```

paulromano commented on docs/source/usersguide/depletion.rst:
```diff
@@ -56,11 +36,40 @@ time::
-Note that the coupling between the transport solver and the transmutation solver
-happens in-memory rather than by reading/writing files on disk.
+Note that the coupling between the reaction rate solver and the transmutation
+solver happens in-memory rather than by reading/writing files on disk. OpenMC has two categories of transport operators for obtaining transmutation reaction
```
```suggestion
solver happens in-memory rather than by reading/writing files on disk. OpenMC
has two categories of transport operators for obtaining transmutation reaction
```

paulromano commented on docs/source/usersguide/depletion.rst:
```diff
@@ -177,3 +192,162 @@ across all material instances.
+The :meth:`~openmc.deplete.MicroXS.from_model()` method will produce a
+:class:`~openmc.deplete.MicroXS` object with microscopic cross section data in
+units of ``b``, which is what :class:`~openmc.deplete.IndependentOperator`
+expects the units to be. The :class:`~openmc.deplete.MicroXS` class also includes functions to read in cross section data directly from a ``.csv`` file or from data arrays::
```
```suggestion
expects the units to be. The :class:`~openmc.deplete.MicroXS` class also
includes functions to read in cross section data directly from a ``.csv`` file
or from data arrays::
```

paulromano commented on docs/source/usersguide/depletion.rst:
```diff
@@ -177,3 +192,162 @@ across all material instances.
+Transport-independent depletion
+===============================
+
+.. note::
```
Maybe change this to warning?
```suggestion
.. warning::
```

paulromano commented on docs/source/usersguide/depletion.rst:
```diff
@@ -177,3 +192,162 @@ across all material instances.
+    micro_xs = openmc.deplete.MicroXS()
+    ...
+
+    op = IndependentOperator(materials, micro_xs, chain_file)
```
```suggestion
    op = openmc.deplete.IndependentOperator(materials, micro_xs, chain_file)
```

paulromano commented on docs/source/usersguide/depletion.rst:
```diff
@@ -177,3 +192,162 @@ across all material instances.
+
+The :meth:`~openmc.deplete.MicroXS.from_model()` method will produce a
+:class:`~openmc.deplete.MicroXS` object with microscopic cross section data in
+units of ``b``, which is what :class:`~openmc.deplete.IndependentOperator`
```
```suggestion
units of barns, which is what :class:`~openmc.deplete.IndependentOperator`
```

paulromano commented on docs/source/pythonapi/deplete.rst:
```diff
@@ -12,19 +12,20 @@ Primary API
 are:
 
-    1) A transport operator
+    1) A transpor operator
```
```suggestion
    1) A transport operator
```

paulromano commented on openmc/deplete/abc.py:
```diff
@@ -79,11 +78,12 @@ def change_directory(output_dir):
     """Abstract class defining a transport operator
 
-    Each depletion integrator is written to work with a generic transport
+    Each depletion integrator is written to work with a generic depletion
```
```suggestion
    Each depletion integrator is written to work with a generic transport
```

paulromano commented on openmc/deplete/abc.py:
```diff
@@ -598,7 +604,7 @@ def __init__(self, operator, timesteps, power=None, power_density=None,
                 source_rates = [p*operator.heavy_metal for p in power_density]
         elif source_rates is None:
-            raise ValueError("Either power, power_density, or source_rates must be set")
+            raise ValueError("Either power, power_density, source_rates must be set")
```
```suggestion
            raise ValueError("Either power, power_density, or source_rates must be set")
```

paulromano commented on openmc/deplete/helpers.py:
```diff
@@ -610,7 +615,7 @@ class FissionYieldCutoffHelper(TalliedFissionYieldHelper):
     fast_energy : float, optional
         Energy of yield data corresponding to fast yields.
-        Default: 500 [kev]
+        Default: 500 [KeV]
```
Default is in the signature so let's just remove it
```suggestion
```

paulromano reviewed (changes requested):
I think we're almost there. By the way -- I'm not sure if you're aware but with suggested changes, you can go to the "Files changed" tab, add them to a batch and commit all in one go without having to make the changes yourself manually. I would also consider it a best practice since the person who makes the suggested changes gets attribution on the commit as a coauthor.

yardasol commented:
I didn't know that, thanks for the tip!
> I would also consider it a best practice since the person who makes the suggested changes gets attribution on the commit as a coauthor.

Whoops! Definitely good to know this for the future. Don't want to be stealing anyone's work.

yardasol commented:
@paulromano Please hold off on merging. I have a small change I'd like to propose that will make using the `MicroXS.from_model()` method  bit more useful and practical. Right now it will only return cross sections from nuclides that are present in the `material` contained in the model, but a user would realistically have a chain file that contains hundreds and thousands of nuclides, not all of which are going to be present in their initial material.

paulromano commented:
@yardasol If it's not too difficult, let's go ahead and include that update with this PR. I agree that, in light of those arguments, the `MicroXS.from_model` class wouldn't be terribly useful.

yardasol commented:
Okay, I've implemented this feature using `dilute_initial` in an identical way to how it's used in `CoupledOperator`. I've also added `chain_file` as a positional argument to `from_model`

paulromano commented on docs/source/usersguide/depletion.rst:
```diff
@@ -177,3 +193,166 @@ across all material instances.
+    materials = openmc.Materials()
+    ...
+
+    micro_xs = openmc.deplete.MicroXS
```
Probably should add `from_...(...)`?

paulromano commented on openmc/deplete/microxs.py:
```diff
@@ -0,0 +1,229 @@
+        return reactions, diluted_materials
+
+    @staticmethod
+    def _get_nuclides_with_data(cross_sections):
+        """Loads cross_sections.xml file to find nuclides with neutron data
+
+        Parameters
+        ----------
+        cross_sections : str
+            Path to cross_sections.xml file
+
+        Returns
+        -------
+        nuclides : set of str
+            Set of nuclide names that have cross secton data
+
+        """
+        nuclides = set()
+        data_lib = DataLibrary.from_xml(cross_sections)
+        for library in data_lib.libraries:
+            if library['type'] != 'neutron':
+                continue
+            for name in library['materials']:
+                if name not in nuclides:
+                    nuclides.add(name)
+
+        return nuclides
```
This is exactly the same as `CoupledOperator._get_nuclides_with_data`. Let's just make this a normal function and then call it from both the `CoupledOperator` and `MicroXS` classes.

yardasol replied:
done

paulromano commented on openmc/deplete/microxs.py:
```diff
@@ -0,0 +1,229 @@
+        return micro_xs
+
+    @classmethod
+    def _add_dilute_nuclides(cls, chain_file, model, dilute_initial):
```
Add a docstring on this -- obviously not important for users since this is internal but docstrings are also useful for us developers :smile:

yardasol replied:
Fair point

paulromano commented on openmc/deplete/microxs.py:
```diff
@@ -0,0 +1,229 @@
+        reactions, diluted_materials = cls._add_dilute_nuclides(chain_file,
+                                                                model,
+                                                                dilute_initial)
+        diluted_materials.export_to_xml('diluted_materials.xml')
```
Doesn't look like this is used
```suggestion
```

yardasol replied:
This was for testing purposed and forgot to remove it when I committed.

paulromano commented on openmc/deplete/microxs.py:
```diff
@@ -0,0 +1,229 @@
+        for material in model.materials:
+            if material.depletable:
+                nuc_densities = material.get_nuclide_atom_densities()
+                dilute_density = 1.E-24 * dilute_initial
```
To match prevailing style throughout OpenMC codebase
```suggestion
                dilute_density = 1.e-24 * dilute_initial
```

yardasol replied:
In that case there are a few other places this needs to be updated.

paulromano commented on openmc/deplete/microxs.py:
```diff
@@ -0,0 +1,229 @@
+            if material.depletable:
+                nuc_densities = material.get_nuclide_atom_densities()
+                dilute_density = 1.E-24 * dilute_initial
+                material.set_density('sum')
+                for nuc, density in nuc_densities.items():
+                    material.remove_nuclide(nuc)
+                    material.add_nuclide(nuc, density)
+                for burn_nuc in burnable_nucs:
+                    if burn_nuc not in nuc_densities:
+                        material.add_nuclide(burn_nuc,
+                                             dilute_density)
+            diluted_materials.append(material)
```
This seems like it could be a potentially annoying side effect of calling `MicroXS.from_model`. Namely, the `Model` object that's passed in will be changed as a result of this block of code. Probably what we should do is 1) create new `Material` objects that match the original ones but have the dilute nuclides added, 2) create a new copy of `Model` that uses the same geometry but the new materials.

yardasol replied:
I attempted implementing this but am now getting a weird error:

```python
    def test_from_model(model):
        ref_xs = MicroXS.from_csv('test_reference.csv')
>       test_xs = MicroXS.from_model(model, model.materials[0], CHAIN_FILE)

/home/ooblack/projects/openmc/tests/regression_tests/microxs/test.py:50: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/home/ooblack/projects/openmc/openmc/deplete/microxs.py:98: in from_model
    df = xs[rxn].get_pandas_dataframe(xs_type='micro')
/home/ooblack/projects/openmc/openmc/mgxs/mgxs.py:2112: in get_pandas_dataframe
    df['mean'] /= np.tile(densities, tile_factor)
/home/ooblack/miniconda3/envs/openmc-env/lib/python3.10/site-packages/pandas/core/generic.py:11387: in __itruediv__
    return self._inplace_method(
/home/ooblack/miniconda3/envs/openmc-env/lib/python3.10/site-packages/pandas/core/generic.py:11352: in _inplace_method
    result = op(self, other)
/home/ooblack/miniconda3/envs/openmc-env/lib/python3.10/site-packages/pandas/core/ops/common.py:70: in new_method
    return method(self, other)
/home/ooblack/miniconda3/envs/openmc-env/lib/python3.10/site-packages/pandas/core/arraylike.py:124: in __truediv__
    return self._arith_method(other, operator.truediv)
/home/ooblack/miniconda3/envs/openmc-env/lib/python3.10/site-packages/pandas/core/series.py:5636: in _arith_method
    return base.IndexOpsMixin._arith_method(self, other, op)
/home/ooblack/miniconda3/envs/openmc-env/lib/python3.10/site-packages/pandas/core/base.py:1295: in _arith_method
    result = ops.arithmetic_op(lvalues, rvalues, op)
/home/ooblack/miniconda3/envs/openmc-env/lib/python3.10/site-packages/pandas/core/ops/array_ops.py:222: in arithmetic_op
    res_values = _na_arithmetic_op(left, right, op)
/home/ooblack/miniconda3/envs/openmc-env/lib/python3.10/site-packages/pandas/core/ops/array_ops.py:163: in _na_arithmetic_op
    result = func(left, right)
/home/ooblack/miniconda3/envs/openmc-env/lib/python3.10/site-packages/pandas/core/computation/expressions.py:239: in evaluate
    return _evaluate(op, op_str, a, b)  # type: ignore[misc]
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

op = <built-in function truediv>, op_str = '/'
a = array([1.98362959e-04, 1.04605679e-02, 1.92481982e-02, 3.95592500e-05,
       3.47830212e-06, 7.17484119e-09])
b = array([], dtype=float64)

    def _evaluate_standard(op, op_str, a, b):
        """
        Standard evaluation.
        """
        if _TEST_MODE:
            _store_test_result(False)
>       return op(a, b)
E       ValueError: operands could not be broadcast together with shapes (6,) (0,)

/home/ooblack/miniconda3/envs/openmc-env/lib/python3.10/site-packages/pandas/core/computation/expressions.py:69: ValueError

```

It's unclear how my change is effecting the `mgxs` side of things. I'll poke around tomorrow to see if I can fix it, but putting it here in the meantime in case anyone want's to take a crack at it.

paulromano replied:
I think the problem may be something like the following:
- `reaction_domain` is set to `model.materials[0]`
- Under the hood, we change model.materials to the new materials with dilute nuclides
- `reaction_domain` is passed to `FissionXS`, which then tries to determine the nuclides based on that material (i.e., `model.materials[0]`, not the new material with dilute nuclides)
- Thus, the tallies that are created don't include all the nuclides we need

yardasol replied:
subtle point! I'll keep this in mind.

yardasol replied:
So my workaround for this is just to create a `deepcopy` of the original materials object, do all the dilution to the `Materials` object like I did before, and then once the openmc simulation is done, set `model.materials` to the deepcopy.

paulromano commented on openmc/deplete/microxs.py:
```diff
@@ -0,0 +1,229 @@
+        reactions = chain.reactions
+        cross_sections = _find_cross_sections(model)
+        nuclides_with_data = cls._get_nuclides_with_data(cross_sections)
+        burnable_nucs = [nuc.name for nuc in chain.nuclides
+                               if nuc.name in nuclides_with_data]
```
```suggestion
        burnable_nucs = [nuc.name for nuc in chain.nuclides
                         if nuc.name in nuclides_with_data]
```

paulromano commented on openmc/deplete/microxs.py:
```diff
@@ -0,0 +1,235 @@
+        burnable_nucs = [nuc.name for nuc in chain.nuclides
+                         if nuc.name in nuclides_with_data]
+        diluted_materials = Materials()
+        reset_auto_ids()
+        for material in model.materials:
+            if material.depletable:
+                nuc_densities = material.get_nuclide_atom_densities()
+                dilute_density = 1.e-24 * dilute_initial
+                diluted_material = Material()
```
Rather than use `reset_auto_ids` (which itself is another undesired side effect), let's just silence the warning from creating another material with the same ID:
```Python
import warnings

with warnings.catch_warnings():
    warnings.simplefilter("ignore", openmc.IDWarning)
    diluted_material = Material(material_id=material.id)
...
```

paulromano commented on openmc/deplete/microxs.py:
```diff
@@ -0,0 +1,233 @@
+        model.tallies = original_tallies
+        model.materials = original_materials
+
+        return cls(micro_xs)
```
`micro_xs` was created using `cls` above -- why is it necessary to pass it to `cls` again?

yardasol replied:
We are using `pandas.concat`, which returns a `pandas.DataFrame` instead of a `MicroXS`, so we need initialize the `MicroXS` object again.

paulromano commented on openmc/deplete/microxs.py:
```diff
@@ -0,0 +1,233 @@
+                    xs[rxn].load_from_statepoint(sp)
+
+        # Build the DataFrame
+        micro_xs = cls()
+        for rxn in xs:
+            df = xs[rxn].get_pandas_dataframe(xs_type='micro')
+            df.index = df['nuclide']
+            df.drop(['nuclide', xs[rxn].domain_type, 'group in', 'std. dev.'], axis=1, inplace=True)
+            df.rename({'mean': rxn}, axis=1, inplace=True)
+            micro_xs = concat([micro_xs, df], axis=1)
+
+        # Revert to the original tallies and materials
+        model.tallies = original_tallies
+        model.materials = original_materials
+
+        return cls(micro_xs)
```
Here's a nice simplification to make -- this pulls out the 'mean' column as a `Series`, puts it in a dictionary, and then constructs the dataframe from the dict of series. Also, our standard abbreviation in OpenMC for "reaction" is `rx`
```suggestion
        series = {}
        for rx in xs:
            df = xs[rx].get_pandas_dataframe(xs_type='micro')
            series[rx] = df.set_index('nuclide')['mean']

        # Revert to the original tallies and materials
        model.tallies = original_tallies
        model.materials = original_materials

        return cls(series)
```

yardasol commented:
@paulromano I think this should be okay to merge now, unless anyone has any further comments.

paulromano reviewed (approved):
Thanks for getting this to the finish line @yardasol!
